rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,361 @@
+
+- `username` (string): the username of the logged-in user
+- `badgeNumber` (string, optional): the badge number of the logged-in user
+- `stop` (string, optional): GTFS stop ID at which the logged-in user is located
+```json
+{
+  "username": "example",
+  "badgeNumber": "123456",
+  "stop": "place-lake"
+}
+
+or
+
+{
+  "username": "pswartz"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Car
+Cars (individual parts of a train consist) will be described by their 4 digit number.
+```json
+{
+  "carNumber": "3800"
+}
+```
+
+### Operator
+Operators will be described by their badge number:
+```json
+{
+  "badgeNumber": "123"
+}
+```
+
+### Optional T
+A value that can either be an instance of T, or `null`. For example, Optional Car can be
+```json
+{ "carNumber": "3800"}
+# or 
+null
+```
+
+### Time
+A [GTFS Time](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#field-types).
+> Time in the HH:MM:SS format (H:MM:SS is also accepted). The time is measured from "noon minus 12h" of the service day (effectively midnight except for days on which daylight savings time changes occur). For times occurring after midnight, enter the time as a value greater than 24:00:00 in HH:MM:SS local time for the day on which the trip schedule begins.  
+> _Example: `14:30:00` for 2:30PM or `25:35:00` for 1:35AM on the next day._
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matches a GTFS trip, in which case RTR will us the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+**Added Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `id` (string): unique value representing this trip for the given `serviceDate`
+```json
+`{"serviceDate": "2023-01-20", "id": "ADDED-123"}
+```
+**Scheduled Trip (departure time from origin)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `departureTime` (Time): when the vehicle is scheduled to depart the given stop
+- `startStop` (string): GTFS stop ID the trip is departing from (either `location_type` 0 or 1)
+```json
+`{"serviceDate": "2023-01-20", "departureTime": "25:01:02", "startStop": "place-lake"}
+```
+**Scheduled Trip (arrival time at destination)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `arrivalTime` (Time): when the vehicle is scheduled to arrive at the given stop
+- `endStop` (string): GTFS stop ID the trip is ending at (either `location_type` 0 or 1)
+```json
+`{"serviceDate": "2023-01-20", "arrivalTime": "09:10:11", "endStop": "place-lake"}
+```
+**Scheduled Trip (known GTFS trip ID)
+- `serviceDate` (RFC3999 date): the service date for the trip
+- `gtfsId` (string): GTFS trip ID
```

Comment:
One issue around GTFS trip ids that hasn't really been resolved yet:

What happens during disruptions or other times when GTFS trips don't exactly correspond to HASTUS trips? GTFS trips are more accurate that HASTUS in the trip times that are supposed to run. But GTFS doesn't have run numbers, and probably shouldn't be in charge of deciding which operators/runs should go on each trip in the new schedule.

Someone has to do the job of deciding who to put on each trip. Up to now, that's been done by inspectors when they heavily edit the outdated paper trainsheets. If Glides uses GTFS/ODS data to make the trainsheets, then Glides might have to figure out how to do that merge between HASTUS rating personnel schedules and GTFS trip schedules.

This is something Glides will probably have to do research on before using GTFS/ODS in trainsheets, and therefore will make it harder for Glides to include GTFS ids in these events.

Reply:
@skyqrose I'll let you and @boringcactus decide whether to include this. It was added in response to https://github.com/mbta/technology-docs/pull/18#discussion_r1083049821

Reply:
There are two ways for Glides schedule imports to work:

1. Glides imports schedule data from HASTUS, and does not consider schedule data in GTFS + ODS. This is the current process.
2. Glides imports schedule data from GTFS + ODS, and does not consider schedule data in HASTUS. This is the future process.

My idea with using GTFS trip IDs was that they'd be present _instead of_ the time- and station-based trip keys for scheduled trips, so that RTR wouldn't have to match times to trips at all. As I [said](https://github.com/mbta/technology-docs/pull/18#discussion_r1083049821),
> This would mean the implementation of this RFC would be blocked on Transit Data publishing ODS and Glides loading schedules from GTFS+ODS instead of directly from HASTUS exports, but I suspect it would simplify the implementation in RTR. I don't know how much of an effect that'd actually have, though, so it might or might not be worth it.

I think it would be silly to define both the GTFS-based and non-GTFS-based trip keys here. The only advantage to the GTFS trip keys is that RTR can use them directly, but if it also needs to support non-GTFS trip keys, there's no point.

If turning the non-GTFS-based scheduled trip keys into GTFS trip IDs would be complicated for RTR, and ODS is going to be ready at a reasonable time, then the GTFS-based scheduled trip key should be the only kind of scheduled trip key. If turning non-GTFS-based scheduled trip keys into GTFS trip IDs would be simple for RTR, and/or ODS isn't going to be ready anytime soon, then the non-GTFS-based scheduled trip keys should be the only kinds of scheduled trip key.

Reply:
This raises to me the important question of what ODS should look like when we put in planned disruptions into GTFS…but that’s outside the scope of this particular RFC. 🙂😬

Reply:
Given the confusion here, I'm going to pull the GTFS ID out. The `scheduledTime` values work with both HASTUS and ODS, but the GTFS ID only works with ODS (maybe).